### PR TITLE
sam: atsamc21: driver for CAN for SAMC21

### PR DESCRIPTION
**Pull request by benys**
Driver was based on can_sam.

Signed-off-by: Kamil Serwus <kserwus@gmail.com>

### CODE REVIEW

File: drivers/can/can_sam0.c
Comment (henrikbrixandersen) [inline drivers/can/can_sam0.c:37]: Does the SAM0 series only use one M_CAN interrupt line?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
We will need a way of testing this - at least the compilation part. Can this be enabled on a board?

File: drivers/can/can_sam0.c
Comment (benys) [reply drivers/can/can_sam0.c:37]: Yes. SAM21 have only one interrupt line. But I saw that can_mcan is expected 2 lines -> I fire 2 ISRs

**Comment by benys** [comment]
> We will need a way of testing this - at least the compilation part. Can this be enabled on a board?

This CAN works only with SAMCAN21x. I have in my solution connected to board intruduced at #50384 PR. 
In my opionion I should add connection to board/soc after merged #50384? Becouse only C21 implements this SAM0_CAN

BTW In my repo it fully works. You can check my old commit: https://github.com/benys/zephyr/commit/55559a875e6ac83e73927d861b03ecd1ca3ee757

**Comment by henrikbrixandersen** [comment]
Okay. I suggest postponing this to after #50384 is merged, then. That was the initial feedback given in that PR as well.

**Comment by alexanderwachter** [comment]
Please provide a meaningful commit message

**Comment by str4t0m** [comment]
Converted to a draft for now, as board files still need to be added after #50384 was merged.

**Comment by benys** [comment]
@henrikbrixandersen i have a problem, that CAN can send messages (i receive messages by CAN usb stick), but when I send by CAN usb stick - i don't receive interrupt.
I set this filter:
```
    const struct can_filter change_led_filter = {
		.id_type = CAN_EXTENDED_IDENTIFIER,
		.rtr = CAN_DATAFRAME,
		.rtr_mask = 1,
		.id_mask = 0
	};

    ret = can_add_rx_filter_msgq(can_dev, &can_msgq, &change_led_filter);
	if (ret == -ENOSPC) {
		printk("Error, no filter available!\n");
		return;
	}
```
Do you have idea what I have to check?

**Comment by benys** [comment]
@henrikbrixandersen 
> * Have you tried using the CAN shell for testing?
Not yet, but I run sam counter sample. And I extend it a litte bit. On loopback mode it fully works. With external CAN commands it not handle incoming CAN frames.
> * Are the pinctrl settings correct?
I tripled checked. I honestly I port my application which originaly used plain ASF4. So I checked pinctrls by comparison one-to-one. TXpin works becouse it send data. SO i believe that in in zephyr RXpin also is corecrly connected.
> * Is the CAN interrupt enabled correctly?
Yes, but I have idea that problem is that zephyr mcan driver uses two interrupts per device, but SAMC has only one line. I tried investigate, that maybe mcan driver set to use second interrupts for incoming data from external world.
> * Is the CAN interrupt triggered at all?
Yes for example:
```
LED ON
[00:00:05.004,000] <dbg> can_mcan: can_mcan_send: Sending 2 bytes. Id: 0x12345, ID type: extended   
[00:00:05.004,000] <dbg> can_mcan: can_mcan_send: Waiting for TX complete
[00:00:05.005,000] <dbg> can_mcan: can_mcan_line_1_isr: RX FIFO0 INT
[00:00:05.005,000] <dbg> can_mcan: can_mcan_get_message: Frame on filter 28, ID: 0x12345
Counter received: 2
[00:00:06.005,000] <dbg> can_mcan: can_mcan_send: Sending 1 bytes. Id: 0x10, ID type: standard   
[00:00:06.006,000] <dbg> can_mcan: can_mcan_line_1_isr: RX FIFO0 INT
[00:00:06.006,000] <dbg> can_mcan: can_mcan_get_message: Frame on filter 0, ID: 0x10
```
> * Have you tried using a debugger to examine what happens?
Yes :-) without debugger it is inposible to investigate :)

I tried check with second interupt for bosch IP. If you have some ideas please give me :)

**Comment by alexanderwachter** [comment]
Looks like the loopback mode works.
That means it is likely that it is either pinctrl, or a problem wit the bus (baudrate, termination, issue with the transceiver, ...)

**Comment by benys** [comment]
@henrikbrixandersen  and @alexanderwachter  i think that i found I receive some of messages.
I forget that "if you work with can you must found mcan datasheet" :)
```
#ifdef CONFIG_CAN_STM32FD
	can->ils = CAN_MCAN_ILS_RXFIFO0 | CAN_MCAN_ILS_RXFIFO1;
#else
	can->ils = 0;//CAN_MCAN_ILS_RF0N | CAN_MCAN_ILS_RF1N;
#endif
	can->ile = CAN_MCAN_ILE_EINT0; //| CAN_MCAN_ILE_EINT1;
```
in mcan driver was implemented that some interrupts was handled by second line. 
So I have a question - you recomend add elif CONFIG_CAN_SAM0 ? with fixed declaration for SAMC?

BTW i need now investiage why have many bus_off state.

**Comment by henrikbrixandersen** [comment]
> So I have a question - you recomend add elif CONFIG_CAN_SAM0 ?

You can add that for now. I am working on getting rid of all platform specific code in the M_CAN backend driver, though.

**Comment by benys** [comment]
@henrikbrixandersen @alexanderwachter i had last problem (i hope). NBTP computation.
I discovered that zephry compute this registry as:

**generation from zephyr**
- ntseg2=3
- ntseg1=26
- nbrp=0

so 

- tq = 1
- ntseq1+ntseg2+3=32
- product = 32*1=32

**my original sample**
- ntseg2=0
- ntseg1=1
- nbrp=7

so 
- tq = 8
- ntseq1+ntseg2+3=4
- product =4*8=32

But first sample - produces a lot of receive errors (counter is not zero) and my second samples works very well.
Did you have same problems in the past? How can I handle this problem?

**Comment by alexanderwachter** [comment]
Well, you can set the sample point a bit earlier to meet your old values. However, the sample points used by Zephyr are recommended by the CiA

**Comment by benys** [comment]
@alexanderwachter thank you.
I used to validate this harware: https://ucandevices.github.io/uccb.html so I understand that it propably set different params a litte not compatible".

So I push my changes related with only 1 interrupt line (instead of two).

**Comment by benys** [comment]
I added changes related with bindings/boards to this PR.
For example in my board "counter" sample works.

**Comment by henrikbrixandersen** [comment]
> I added changes related with bindings/boards to this PR.
> For example in my board "counter" sample works.

Your board needs to pass all the tests/samples tagged with `can`. Please run `twister --device-testing -t can ...` against your board and confirm that all tests are passing (you may or may not run into trouble with the ISO-TP tests, please report back if you do).

Also, please split these changes into multiple, logically separate commits. One for adding the compatible, one for adding the driver, one for adding SoC support, board support etc.

**Comment by benys** [comment]
@henrikbrixandersen I seperated commits in logically:
* one for driver and required changed in mcan
* one for soc support
* one for board configuration

Yes the twister passes, I forget mention it :-)

File: drivers/can/can_sam0.c
Comment (henrikbrixandersen) [inline drivers/can/can_sam0.c:55]: SERCOM?

**Review by henrikbrixandersen — DISMISSED** [review]
Looks good! One small observation, not a blocker.

File: drivers/can/can_sam0.c
Comment (str4t0m) [inline drivers/can/can_sam0.c:45]: The divider seems to be only used for rate calculation, configuration of the divider is missing.

File: drivers/can/can_sam0.c
Comment (str4t0m) [inline drivers/can/can_sam0.c:13]: Could you please also add `#include <zephyr/irq.h>` as you are using `IRQ_CONNECT`

**Review by str4t0m — CHANGES_REQUESTED** [review]
A few comments.

**Comment by henrikbrixandersen** [comment]
@str4t0m Please revisit.

**Comment by str4t0m** [comment]
There are quite a few minor issues, so it would be great if @benys could address them and rebase on the latest main.

**Comment by benys** [comment]
I will rebase ASAP :) 
In very near feature I have to come back to this PR:) beocuse I need it.

**Comment by stephanosio** [comment]
> I will rebase ASAP :)
> In very near feature I have to come back to this PR:) beocuse I need it.

@benys FYI, the feature merge freeze for v3.3.0 is later today (27th January), so you will need to do it very quick if you want this feature in the upcoming release.

**Comment by benys** [comment]
> > I will rebase ASAP :)
> > In very near feature I have to come back to this PR:) beocuse I need it.
> 
> @benys FYI, the feature merge freeze for v3.3.0 is later today (27th January), so you will need to do it very quick if you want this feature in the upcoming release.

I will try!

**Comment by benys** [comment]
@stephanosio @str4t0m I fixed all issues - I hope :)

File: drivers/can/can_sam0.c
Comment (str4t0m) [inline drivers/can/can_sam0.c:46]: Thanks!
I am not really familiar with Atmel SoCs, but in `can_sam0_clock_enable()` you are setting `GCLK_GENCTRL_SRC_OSC48M` as clock source for the _clock generator 1_, which I assume is a fixed 48MHz oscillator. But here you are using `SOC_ATMEL_SAM0_GCLK0_FREQ_HZ` which depends on the selected sysclock frequency and might be changed. Shouldn't you set it to 48MHz or a macro representing the 48MHz here?

Also, the clock generator 1 is part of the Generic Clock Controller peripheral, so having it's divider as a property of the CAN is not ideal. Maybe @nandojve can weight in? Can we introduce that as a temporary solution until a generic clock controller driver is available for the series?

File: boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts
Comment (henrikbrixandersen) [inline boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:162]:
```suggestion
	bus-speed = <125000>;
```

File: boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts
Comment (henrikbrixandersen) [inline boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:174]:
```suggestion
	bus-speed = <125000>;
```

File: dts/bindings/can/atmel,sam0-can.yaml
Comment (henrikbrixandersen) [inline dts/bindings/can/atmel,sam0-can.yaml:7]: Aren't you missing an include on `bosch,m_can-base.yaml`?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Review by str4t0m — COMMENTED** [review]
Thanks for the update.
Most issues are resolved, but I am not sure about configuring a GCLK channel divider via a property of the CAN node.

**Comment by benys** [comment]
> Thanks for the update.
> Most issues are resolved, but I am not sure about configuring a GCLK channel divider via a property of the CAN node.

I had the same issue. I checked how it was implemented at armel SAM - it was the same pattern.

For now c21 doesn't use GLCK other then 0.

File: boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts
Comment (benys) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:162]: I don't understand :( i changed it as i receive review from other person.

Have i to change it?

Comment (str4t0m) [reply boards/arm/atsamc21n_xpro/atsamc21n_xpro.dts:162]: Yes please. The only data phase(CAN-FD) should have the higher bit rate of 1MHz.

File: dts/bindings/can/atmel,sam0-can.yaml
Comment (benys) [reply dts/bindings/can/atmel,sam0-can.yaml:7]: In my opinioin - no - because `bosch,m_can-base.yaml` is defined on controller level, not in individual module.
The same way is defined for sam-can.

File: drivers/can/can_sam0.c
Comment (nandojve) [reply drivers/can/can_sam0.c:46]: I'm not expert on CAN but as far I know clock should be very accurate. I think there is no problem to use `OSC48M` as base clock with proper factory (MCHP) calibration. What could be an issue is that DPLL is not enabled at this moment and that, on my view, should be the GCLKGEN reference for CAN. Usually a `XOSC[32K]` are a better reference for DPLL but you can use `OSC48M` -> `GLKC` -> `DPLL input ref` that will be far better than only `GLKC` (in theory).

Usually without a good clock base CAN may suffer when temperature changes, for instance.

File: drivers/can/can_sam0.c
Comment (nandojve) [inline drivers/can/can_sam0.c:56]:
```diff
+	GCLK->GENCTRL[1].reg = GCLK_GENCTRL_SRC(GCLK_GENCTRL_SRC_OSC48M)
+			     | GCLK_GENCTRL_DIV(cfg->divider)
+			     | GCLK_GENCTRL_GENEN;
```

File: drivers/can/can_sam0.c
Comment (nandojve) [inline drivers/can/can_sam0.c:60]:
```diff
+	GCLK->PCHCTRL[cfg->gclk_core_id].reg = GCLK_PCHCTRL_GEN_GCLK1
+					     | GCLK_PCHCTRL_CHEN;
```

File: drivers/can/can_sam0.c
Comment (nandojve) [inline drivers/can/can_sam0.c:113]: Keep style...fix all entries

```diff
-		},
-	.timing_max = {
+	},
+	.timing_max = {
```

File: drivers/can/can_sam0.c
Comment (nandojve) [inline drivers/can/can_sam0.c:185]: Keep `\` vertically align.
Use only TABs for this purpose
Loop longest line and add 1 TAB+\
Then repeat for the others.
Zephyr is 8 space TAB like Linux.

```diff
#define SAM0_CAN_IRQ_CONNECT(inst, m)							\
...
+		.gclk_core_id = DT_INST_CLOCKS_CELL_BY_NAME(inst, gclk, periph_ch),	\
		.divider = DT_INST_PROP(inst, divider),					\
		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(inst),				\
		.config_irq = config_can_##inst##_irq,					\
	};										\
...
#define CAN_SAM0_DEVICE_INST(inst)							\
	DEVICE_DT_INST_DEFINE(inst, &can_sam0_init, NULL,				\
			      &can_mcan_data_##inst,					\
			      &can_mcan_cfg_##inst,					\
			      POST_KERNEL, CONFIG_CAN_INIT_PRIORITY,			\
			      &can_sam0_driver_api);
...
#define CAN_SAM0_INST(inst)								\
	PINCTRL_DT_INST_DEFINE(inst);							\
	CAN_SAM0_IRQ_CFG_FUNCTION(inst)							\
	CAN_SAM0_CFG_INST(inst)								\
	CAN_SAM0_DATA_INST(inst)							\
```

**Review by nandojve — CHANGES_REQUESTED** [review]
Overall, LGTM. I only checked a few styles that are used for these series.

I let clock decision for CAN folks. I assume that HW works and should be enough to introduce the feature. Maybe a better commit message that advice that clock should be improved to DPLL may help users.

File: drivers/can/can_sam0.c
Comment (benys) [reply drivers/can/can_sam0.c]: I am sorry that forget this tyle. I fixed :)

**Review by nandojve — DISMISSED** [review]

**Comment by nandojve** [comment]
I took liberty to add removed persons from review list because I think it was by mistake.

**Comment by benys** [comment]
> I took liberty to add removed persons from review list because I think it was by mistake.

Yes it was by mistake. I didn't know how revert my mistake.

File: drivers/can/can_sam0.c
Comment (henrikbrixandersen) [reply drivers/can/can_sam0.c:46]: What is recommended clock source/frequency for the MCAN block according to the datasheet?

Comment (benys) [reply drivers/can/can_sam0.c:46]: @henrikbrixandersen documentation doesn't specify recomended source of clock.

> Clocks
> An AHB clock (CLK_CAN_AHB) is required to clock the CAN. This clock can be configured in the Main Clock
peripheral (MCLK) before using the CAN, and the default state of CLK_CAN_AHB can be found in the
MCLK.AHBMASK register.
> A generic clock (GCLK_CAN) is required to clock the CAN. This clock must be configured and enabled in the generic
clock controller before using the CAN.
> This generic clock is asynchronous to the bus clock (CLK_CAN_AHB). Due to this asynchronicity, writes to certain

Datasheet only specify that MCAN has to be connected to GLCK.

**Comment by benys** [comment]
@henrikbrixandersen I would like to back to this this PR :)
Should I rebase to newest one? Need I do more?

File: drivers/can/can_sam0.c
Comment (str4t0m) [reply drivers/can/can_sam0.c:46]: I my first comment I wanted to highlight that:
- You are using GLCK1 as clock source for the CAN peripheral.
- For rate calculation in line 46 you use GLCK0 frequency divided by the GLCK1 divider
- But GLCK1 is feed by OSC48M directly and not in series after GCLK0.
- This will work now as the system clock(GLCK0) is also fed by OSC48M and not divided. But as soon as that changes, chances are high that the CAN driver is overlooked and rate calculation is wrong.
- But if Atmel maintainers are fine with it I won't block.

**Review by str4t0m — CHANGES_REQUESTED** [review]
Yes please rebase on the latest main.

File: drivers/can/can_sam0.c
Comment (henrikbrixandersen) [reply drivers/can/can_sam0.c:46]: @nandojve Just to make sure - are you okay with the above limitations?

File: drivers/can/can_sam0.c
Comment (henrikbrixandersen) [inline drivers/can/can_sam0.c:76]: Please either add logging of failures or remove the log declaration. I would prefer the former.

**Review by henrikbrixandersen — COMMENTED** [review]
LGTM, pending response from @nandojve.

**Review by nandojve — APPROVED** [review]
LGTM!

File: drivers/can/can_sam0.c
Comment (attie-argentum) [reply drivers/can/can_sam0.c:46]: I'm also not convinced that `SOC_ATMEL_SAM0_GCLK0_FREQ_HZ` is appropriate here...

File: drivers/can/can_sam0.c
Comment (attie-argentum) [inline drivers/can/can_sam0.c:56]: All other SAM parts put this level of clock configuration in the SOC's code, e.g: [`.../soc_samc2x.c`](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/atmel_sam0/common/soc_samc2x.c)

Given that `cfg->divider` comes from each channel in the device tree, I'm also concerned that in applications that use both channels, there could be a silent incompatibility between the two... the device tree currently suggests that these dividers are independent, when they aren't!

I note that the f<sub>GCLK_CANn</sub> is listed as "_48 MHz max_", and the prescaler in the CAN peripheral is fairly limited, so is probably required...

I'm concerned that placing the GCLK1 configuration in the driver will make it less visible and potentially introduce issues in the future...
- At minimum, please add a comment to `soc_samc2x.c` to indicate that GCLK1 is reserved for CAN
- My preference would be to move configuration of GCLK1 into `soc_samc2x.c`
- Consider using two GCLK channels (one for each CAN interface), or perhaps move the `divider` value up into the `can` / `bosch,m_can-base` node

e.g:
https://github.com/zephyrproject-rtos/zephyr/blob/ed2c25effccc55e459e2b96f389d9b1a177daf7d/soc/arm/atmel_sam0/common/soc_saml2x.c#L26-L29

**Review by attie-argentum — DISMISSED** [review]
Thanks @benys ... aside from the small stylistic one, I have concerns about the clock configuration. Please see my comments.

**Review by nandojve — CHANGES_REQUESTED** [review]

File: drivers/can/can_sam0.c
Comment (attie-argentum) [reply drivers/can/can_sam0.c:56]: Ah, well found... Using `GCLK0` @ 48 MHz sounds sensible to me...

File: drivers/can/can_sam0.c
Comment (benys) [reply drivers/can/can_sam0.c]: I changed reference from `SOC_ATMEL_SAM0_GCLK0_FREQ_HZ` to `SOC_ATMEL_SAM0_OSC48M_FREQ_HZ`
I hope that it is ok for you :)

File: drivers/can/can_sam0.c
Comment (benys) [reply drivers/can/can_sam0.c:56]: @attie-argentum  @nandojve 
About clocks I inspired with can_sam and can_mcan.
In my project I need divider to set valid ntseg (i had to set specific ntseg1/ntseg2 values for specific clock speed).

Can I change can0 to config GLCK7 instead of GLCK1? It will be ok for you all?

**Comment by benys** [comment]
I see that tests/drivers/uart/uart_async_api doesnt pass. I see that it doesn't have overlay to configure test.
Should I here add overlay file?

File: drivers/can/can_sam0.c
Comment (nandojve) [reply drivers/can/can_sam0.c]: You mean reserve the GCLK[7] for CAN? I don't see problems with that but you need make sure all SAM0 series enable GCLK[7] at soc. Is that what you want to do, right?

Comment (benys) [reply drivers/can/can_sam0.c]: @nandojve as you can see I "dynamicly" enable at cansam0 driver. So that it isn't need to to enable at soc.
BTW In my knowlage CAN module for sam0 has only samC21 series.

Comment (nandojve) [reply drivers/can/can_sam0.c]: Ah ok! But as a remind that ATSAME5x have CAN-FD too.
Fine for me if you want move from GCLK1 to GCLK7.

Comment (benys) [reply drivers/can/can_sam0.c]: I believe that "conversation" can be resolved :-)
Thanks for info about E5x, I didn't know.

**Comment by nandojve** [comment]
> I see that tests/drivers/uart/uart_async_api doesnt pass. I see that it doesn't have exluding for this board...
>
> What next :-)?

I think there is no difference on drivers so maybe you can just exclude it. There is no point to build the exactly same driver if there is no changes between device series that it is important catch. You can achieve that creating another commit and put it as the first on this series. So, you fix the build problem and introduce new driver.

**Comment by benys** [comment]
> > I see that tests/drivers/uart/uart_async_api doesnt pass. I see that it doesn't have exluding for this board...
> > What next :-)?
> 
> I think there is no difference on drivers so maybe you can just exclude it. There is no point to build the exactly same driver if there is no changes between device series that it is important catch. You can achieve that creating another commit and put it as the first on this series. So, you fix the build problem and introduce new driver.

I created new PR - https://github.com/zephyrproject-rtos/zephyr/pull/56487 but on my computer builds normally. In twister I see that it try to use overlay not from board but only for seggerrtt

**Review by nandojve — DISMISSED** [review]

**Review by henrikbrixandersen — DISMISSED** [review]

**Review by alexanderwachter — DISMISSED** [review]

**Comment by benys** [comment]
@attie-argentum & @str4t0m can you checkPR :-)?
I hope that is OK for you!

**Review by str4t0m — DISMISSED** [review]
Thanks for addressing my concerns.
One note but not blocking.

**Comment by henrikbrixandersen** [comment]
@attie-argentum Please revisit.

**Comment by silverv** [comment]
Is it possible to use CAN1 and were there tests for it?

**Comment by attie-argentum** [comment]
> @attie-argentum Please revisit.

@henrikbrixandersen - apologies for my delay. I've dismissed my review - it looks like my comments have been resolved, and I don't have the time to formally re-review right now. Others have approved, so please continue!

**Comment by fabiobaltieri** [comment]
@benys can you rebase please?

**Comment by benys** [comment]
> @benys can you rebase please?

I rabased, but I will it test +- tommorow at the office :) I will inform you.

**Comment by silverv** [comment]
My experience so far is that with commit 81afde13448808f71b12c4bacc18d9ef4f6649ec, both CAN0 and CAN1 initialized but CAN1 gets stuck at send on a atsamc21g18a. As of [782c99f](https://github.com/zephyrproject-rtos/zephyr/pull/50408/commits/782c99fbdea65453b8173d9db1b787891b06fbba), atsamc21n18a, neither CAN peripheral would get past mcan initialization, I will double check everything now.

File: drivers/can/can_sam0.c
Comment (henrikbrixandersen) [inline drivers/can/can_sam0.c:191]:
```suggestion
                .base = (mm_reg_t)DT_INST_REG_ADDR(inst),               \
		.mclk = (volatile uint32_t *)MCLK_MASK_DT_INT_REG_ADDR(inst),		\
```

Comment (henrikbrixandersen) [reply drivers/can/can_sam0.c:191]: Where are you assigning `config->base`?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

**Comment by benys** [comment]
@silverv @henrikbrixandersen  i little fixed my PR, but after rebase still not works :-(
I am still looking reason...

**Review by henrikbrixandersen — APPROVED** [review]
Thanks!

**Review by nandojve — APPROVED** [review]

**Pull request merged**
